swig/csharp/CMakeLists.txt

```diff
@@ -9,226 +15,626 @@ set(GDAL_CSHARP_INSTALL_DIR
+  GDALOverviews.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/GDALOverviews.cs
+  DEPENDS
+  gdal_csharp
+  SYSTEM_DEPENDS
+  OSGeo.GDAL)
+
+gdal_build_csharp_sample(
+  OUTPUT
+  GDALCreateCopy.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/GDALCreateCopy.cs
+  DEPENDS
+  gdal_csharp
+  SYSTEM_DEPENDS
+  OSGeo.GDAL)
+
+gdal_build_csharp_sample(
+  OUTPUT
+  GDALGetHistogram.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/GDALGetHistogram.cs
+  DEPENDS
+  gdal_csharp
+  SYSTEM_DEPENDS
+  OSGeo.GDAL)
+
+gdal_build_csharp_sample(
+  OUTPUT
+  GDALTest.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/GDALTest.cs
+  DEPENDS
+  gdal_csharp
+  SYSTEM_DEPENDS
+  OSGeo.GDAL)
+
+# Custom Target to make the C# wrap Libraries and use SWIG to build the .cs files
+add_custom_target(csharp_interface DEPENDS gdalconst_wrap osr_wrap ogr_wrap gdal_wrap)
+
+# Custom Target to make the c# bindings - will be run as part of ALL but can also be built seperately
+add_custom_target(csharp_binding ALL DEPENDS csharp_interface OSGeo.GDAL.CONST OSGeo.GDAL OSGeo.OGR OSGeo.OSR)
 
+# Custom Target to build the C# bindings and the sample binaries
 add_custom_target(
   csharp_samples
-  DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/ogrinfo.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/createdata.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/OSRTransform.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALRead.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALReadDirect.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALAdjustContrast.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALDatasetRasterIO.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALWrite.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALDatasetWrite.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALColorTable.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/WKT2WKB.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/OGRGEOS.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/ReadXML.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALInfo.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALOverviews.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALCreateCopy.exe
-          ${CMAKE_CURRENT_BINARY_DIR}/GDALGetHistogram.exe)
-add_custom_target(
-  csharp_binding ALL
-  DEPENDS gdalconst_csharp.dll
-          osr_csharp.dll
-          ogr_csharp.dll
-          gdal_csharp.dll
-          gdalconst_wrap
-          osr_wrap
-          ogr_wrap
-          gdal_wrap
-          csharp_samples)
+  DEPENDS csharp_binding
+          OgrInfo.exe
+          CreateData.exe
+          OSRTransform.exe
+          GDALRead.exe
+          GDALReadDirect.exe
+          GDALAdjustContrast.exe
+          GDALDatasetRasterIO.exe
+          GDALWrite.exe
+          GDALDatasetWrite.exe
+          GDALColorTable.exe
+          WKT2WKB.exe
+          OGRGEOS.exe
+          ReadXML.exe
+          GDALInfo.exe
+          GDALOverviews.exe
+          GDALCreateCopy.exe
+          GDALGetHistogram.exe
+          GDALTest.exe)
 
-if (CSHARP_INTERPRETER OR WIN32)
+# set up the tests
+if (CSHARP_INTERPRETER
```

Comment:
parentheses around member of the OR expression would help: (CSHARP_INTERPRETER OR WIN32) AND NOT GDAL_CSHARP_ONLY